apis/metal3.io/v1alpha1/hostfirmwarecomponents_types.go

```diff
@@ -0,0 +1,118 @@
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
+
+// FirmwareUpdate defines a firmware update specification.
+type FirmwareUpdate struct {
+	Component string `json:"component"`
+	URL       string `json:"url"`
+}
+
+// FirmwareComponentStatus defines the status of a firmware component.
+type FirmwareComponentStatus struct {
+	Component          string      `json:"component"`
+	InitialVersion     string      `json:"initialVersion"`
+	CurrentVersion     string      `json:"currentVersion"`
+	LastVersionFlashed string      `json:"lastVersionFlashed"`
+	UpdatedAt          metav1.Time `json:"updatedAt"`
+}
+
+type UpdatesConditionType string
+
+const (
+	// Indicates that the updates in the Spec are different than Status
+	HostFirmwareComponentsChangeDetected UpdatesConditionType = "ChangeDetected"
+
+	// Indicates if the updates are valid and can be configured on the host
+	HostFirmwareComponentsValid UpdatesConditionType = "Valid"
+)
+
+// HostFirmwareComponentsSpec defines the desired state of HostFirmwareComponents
+type HostFirmwareComponentsSpec struct {
+	Updates []FirmwareUpdate `json:"updates"`
+}
+
+// HostFirmwareComponentsStatus defines the observed state of HostFirmwareComponents
+type HostFirmwareComponentsStatus struct {
+	// Updates is the list of all firmware components that should be updated
+	// they are specified via name and url fields.
+	Updates []FirmwareUpdate `json:"updates"`
+
+	// Components is the list of all available firmware components and their information.
+	Components []FirmwareComponentStatus `json:"components"`
+
+	// Time that the status was last updated
+	// +optional
+	LastUpdated *metav1.Time `json:"lastUpdated,omitempty"`
+
+	// Track whether updates stored in the spec are valid based on the schema
+	// +patchMergeKey=type
+	// +patchStrategy=merge
+	// +listType=map
+	// +listMapKey=type
+	// +optional
+	Conditions []metav1.Condition `json:"conditions,omitempty" patchStrategy:"merge" patchMergeKey:"type" protobuf:"bytes,1,rep,name=conditions"`
+}
+
+//+kubebuilder:object:root=true
+//+kubebuilder:resource:shortName=hfc
+//+kubebuilder:subresource:status
+
+// HostFirmwareComponents is the Schema for the hostfirmwarecomponents API
+type HostFirmwareComponents struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   HostFirmwareComponentsSpec   `json:"spec,omitempty"`
+	Status HostFirmwareComponentsStatus `json:"status,omitempty"`
+}
+
+//+kubebuilder:object:root=true
+
+// HostFirmwareComponentsList contains a list of HostFirmwareComponents
+type HostFirmwareComponentsList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []HostFirmwareComponents `json:"items"`
+}
+
+// Check whether the updates's names are valid
+func (host *HostFirmwareComponents) ValidateHostFirmwareComponents() error {
```

Comment:
Idea for a follow-up: move this to the driver level. Since this is, in fact, driver specific.